Source/DafnyCore/Verifier/Translator.cs

```diff
@@ -3851,6 +3851,122 @@ public enum StmtType { NONE, ASSERT, ASSUME };
       MakeAssert(tok, q, desc, kv);
     }
 
+    public static bool IsExpressionAlways(Expression q, bool truth) {
```

Comment:
Could you move this to VerificationDiagnosticsParams? This file is already so large and this method isn't used here.

Reply:
> Could you move this another file? This file is already so large and this method isn't used here. I think I would prefer it close to the other gutter code, or if you think this is useful for other scenarios maybe in its own file in DafnyCore.

I think it will be interesting to other functions in DafnyCore down the road, so I moved it to its own file `QuickVerifier`.

> Should you ignore assume statements that have {:axiom} on them?

Yes, the same way the compiler ignores them.

> Is there a reason to ignore trivially true assume statements? Is there a use-case for them? It seems simpler to 'warn' about all these assume statements as well.

I've seen the use of `assume {:split_here} true;` so that it triggers Boogie splits of verification without introducing new assertions to verify. The assumptions don't have any reason to be warned about, even if it is simpler. Of course, this is was not for our own Dafny compiler.

> I think @atomb is planning to add an analysis that detects which requires/ensures are always false, using the verifier. Would that be a more powerful version of this?

Yes that will be a more powerful version of this. This is why I renamed the file `QuickVerifier`, because it can say if something is `unsat` or `unknown` for a very small subset of all propositions, but which were relevant for many of our customers. This heuristic feature executes linearly with the size of the conditions and only detect patterns, so it won't slow down verification like @atomb might do - but that will be needed anyway probably at least for CI builds.

Reply:
> I've seen the use of assume {:split_here} true; so that it triggers Boogie splits of verification without introducing new assertions to verify. The assumptions don't have any reason to be warned about, even if it is simpler. Of course, this is was not for our own Dafny compiler.

I'd still be in favor of not adding that edge-case check. I think pushing users to consistently use `assert {:split_here} true;` instead of `assume {:split_here} true` will make the language easier to understand.

Reply:
> I'd still be in favor of not adding that edge-case check. I think pushing users to consistently use assert {:split_here} true; instead of assume {:split_here} true will make the language easier to understand.

Ok another case then. Let's say a user writes this:

`assume a[i] == a[i]` instead of `assume a[i] == a[j];`

If we signal all assumptions, then the first case will be silently ignored as the user will expect the assumption to be underlined as it would be for the second case. It might not help them to spot an issue in the assumption, in that the assumption does not help at all.

So, checking if the assumption is trivially true will give the feedback that, given a user would expect assumptions to be marked, if it's not marked, they will know this assumption is useless.

Reply:
I think Dafny should have the following warnings:
- A warning for every `assume` without {:axiom}. We can incorporate this one in the gutter icons.
- A warning for expressions with free variables that always have the same value, such as `a[i] == a[i]`. For reasons of performance we might only do this check for expressions passed to `assert` and `assume` statements. I suggest we don't incorporate this one in the gutter icons, but just leave it as a warning.

Given the above:
`assume a[i] == a[i]` <-- two warnings
`assume {:axiom} a[i] == a[i]` <-- one warning
`assume a[i] == a[j]` <-- one warning
`assume {:axiom} a[i] == a[j]` <-- zero warnings

It's not yet clear to me what information you want to incorporate into gutter icons. Do you want any warnings that relate to verification to cause the verification context to show up as 'not green' in the gutter?

I can also imagine two other options:
- Any warning, whether it relates to verification or not, causes the gutter icons of the context to show as non-green
- The gutter icons are green as long as no verification errors occur in the context, so all warnings, including those coming from `assume` statements, are ignored.

What I like above the above two options is that I find them simple to understand.

Reply:
Yeah, even though I am planning a deeper analysis for contradictory assumptions I think having a really fast check for obvious cases is worthwhile. Detecting those as early as possible is valuable.

Reply:
Thanks for your feedback, always appreciated. Let's see if we can converge to a practical roadmap.

> I think Dafny should have the following warnings:

Please, consider that the scope of this PR is not about warnings, it's about indicating contextual proof status. You interpreted it as Dafny should add warnings, but I believe this is a different (and useful) point that I want to keep out of scope of this PR. If you want I can work on warnings after this PR. I agree that, for now, we should consider how this PR can interact with warnings reported in the gutter in the future, so that we can choose the "icons" wisely. See the discussion below.

> * A warning for every assume without {:axiom}.

When it will come to warnings, I think you have the right approach. We could just issue two warnings. For this PR, I have no plans of making "double warning gutter icons".

> * A warning for expressions with free variables that always have the same value, such as a[i] == a[i]. For reasons of performance we might only do this check for expressions passed to assert and assume statements.

What you probably had in mind was not "expressions with free variables that always have the same value", but "expressions that are always trivially true", right? Otherwise, you are assuming equality assumptions, which this PR does not.

> 1. Any warning, whether it relates to verification or not, causes the gutter icons of the context to show as non-green

I like the idea of displaying warnings in the gutter in any case in principle - the same way we already display diagnostic errors with triangles - but in this case I would like the gutter to switch from green to the conditional verification symbol **if and only if there is at least one warning that concerns potential global verification issues when the code is verified**.
For example, the warnings that affect verification are nontrivial assumptions not in requires, a forall statement without body, etc.
Warnings that don't affect global verification outcomes, to my opinion, should not trigger the contextual "verified" icons to be changed, such as comparing null to a non-null type, importing an empty module, requires clause not compilable, unnecessary (==) if type parameter is ghost, the conclusion of a forall statement will not be known...

> 2. The gutter icons are green as long as no verification errors occur in the context, so all warnings, including those coming from assume statements, are ignored.

As described in this PR's description, that would defeat the whole purpose of indicating to the user that there are still some assume statements in the scope of the proof. So this is a big no for this one, even if it's simple.

Reply:
> Please, consider that the scope of this PR is not about warnings, it's about indicating contextual proof status. You interpreted it as Dafny should add warnings, but I believe this is a different (and useful) point that I want to keep out of scope of this PR. If you want I can work on warnings after this PR. I agree that, for now, we should consider how this PR can interact with warnings reported in the gutter in the future, so that we can choose the "icons" wisely. See the discussion below.

Actually I'm just talking about warnings to better understand this PR, since they're so related. That's also why I'm arguing to let any `assume` without axiom trigger the warning gutter context, since I'm planning to let any assume without axiom trigger a warning and the two should be consistent. (Ideally we would achieve the consistency by computing (part of) the gutter icons on the client based on the warnings, but now let's do it using the APIs.

> Warnings that don't affect global verification outcomes, to my opinion, should not trigger the contextual "verified" icons to be changed

Maybe that's the right thing to do but why? Why not all warnings? Why would the gutter only try to help you with verification problems instead of anything that requires more actions in that piece of code?

> What you probably had in mind was not "expressions with free variables that always have the same value", but "expressions that are always trivially true", right? Otherwise, you are assuming equality assumptions, which this PR does not.

I was imagining which warnings we'd ideally like to have in the future.

Reply:
> Maybe that's the right thing to do but why? Why not all warnings? Why would the gutter only try to help you with verification problems instead of anything that requires more actions in that piece of code?

I think the primary reason was to make them not too complex, but I would agree it's not a good reason. Let's investigate some more points and see whether we can arrive at something we both like.

Until today, the gutter icons have been used almost exclusively to display information related to code verification, with the only exception of resolution errors. But we display resolution errors using sharp triangles on top of greyed previous verification. We don't use a specific "context" saying there are resolution errors in scope, it can be seen on the entire file on the scroll bar, and locally on the gutter if needed. And I think that's ok.

So while I also want to possibly display warnings on gutter icons (especially when there are no errors or context errors), I would like to not always change the "context" value. I would like to keep true to the original verification gutter icons principle: If the context is one line green, Dafny did not identify something to worry about the relationship between the spec and the code.

Now, I think it's going to be not easy to distinguish warnings that should affect the verification outcome if acted on, and those which don't, but I think we could make an effort. Can we mark the source of such warnings like "Verifier" and "Resolver" ? The first one would impact verification, the second ones wouldn't.
For example, I strongly believe that warnings about dead code, about no trigger for quantifiers found, comparing null to non-null type, should not trigger the feeling that there is something to fix between the spec and the implementation. That would be annoying. They are just there to help optimize code, not to actually warn of something serious that could impact verification. Such warnings could be displayed as a simple icon (like resolution error icons) on top of any verification context that is not an error, not affecting the context itself like verification warnings would.

One the other side, a warning about requires false, assume anything (nontrivial, I would still argue for that) without `{:axiom}`, or while loops without a body, should for sure trigger the verification warning in context.